src/query/sql/src/planner/optimizer/optimizer.rs

```diff
@@ -189,7 +190,8 @@ pub fn optimize_query(
     if enable_distributed_query {
         result = optimize_distributed_query(ctx.clone(), &result)?;
     }
-
+    result = heuristic.optimize_expression(&result, &RESIDUAL_RULES)?;
```

Comment:
Why put it here?

```suggestion
    if ctx.get_settings().get_enable_dphyp()? {
        let (dp_res, optimized) =
            DPhpy::new(ctx.clone(), metadata.clone()).optimize(Arc::new(result.clone()))?;
        if optimized {
            dphyp_optimized = true;
            result = heuristic.optimize_expression(&dp_res, &RESIDUAL_RULES)?;
            result = heuristic.post_optimize(dp_res)?;
        }
    }
```

Reply:
Because we need to consider the `Exchange` operator to avoid the columns required by `Exchange::Hash(Vec<ScalarExpr>)` being pruned.